File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int[] postorder;
    private Map<Integer, Integer> inorderWithIndex;
    private int postorderIndex;

    // LRN
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        this.inorderWithIndex = new HashMap<>();
        this.postorder = postorder;
        this.postorderIndex = postorder.length - 1;
        int index = 0;
        for(int i: inorder) inorderWithIndex.put(i, index++);
        return helper(0, postorder.length - 1);
    }

    private TreeNode helper(int left, int right) {
        if(left > right)
            return null;
        var root = new TreeNode(postorder[postorderIndex--]);
        int rootIndex = this.inorderWithIndex.get(root.val);
        root.right = helper(rootIndex + 1, right);
        root.left = helper(left, rootIndex - 1);
        return root;
    }
}

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

#include "iostream"
#include "map"
using namespace std;

class Solution {
private:    
    int postorderIndex;
    vector<int> postorder;
    map<int, int> inorderIndex;

public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        this->postorder = postorder;
        this->postorderIndex = postorder.size() - 1;
        int inIndex = 0;
        for (int in : inorder) {
            this->inorderIndex[in] = inIndex++;
        }

        return this->helper(0, inorder.size() - 1);
    }

    TreeNode* helper(int left, int right) {
        if (left > right)
            return nullptr;

        TreeNode* root = new TreeNode(this->postorder[postorderIndex--]);
        int rootIndex = this->inorderIndex[root->val];
        root->right = helper(rootIndex + 1, right);
        root->left = helper(left, rootIndex - 1);

        return root;
    }
};